telemetry/internal/appinsights/client_test.go

```diff
@@ -0,0 +1,102 @@
+//go:build goexperiment.synctest
+
+package appinsights
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"io"
+	"log"
+	"net/http"
+	"testing"
+	"testing/synctest"
+)
+
+type fakeTransport struct {
+	code int
+	body []byte
+	err  error
+}
+
+func (t fakeTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	return &http.Response{
+		StatusCode: t.code,
+		Body:       io.NopCloser(bytes.NewBuffer(t.body)),
+	}, t.err
+}
+
+func BenchmarkClientBurstPerformance(b *testing.B) {
+	client := &Client{
+		InstrumentationKey: test_ikey,
+		ErrorLog:           log.New(io.Discard, "", 0),
+		HTTPClient: &http.Client{
+			Transport: fakeTransport{
+				code: http.StatusOK,
+				body: []byte(`{"itemsReceived":4, "itemsAccepted":4, "errors":[]}`),
+			},
+		},
+	}
+
+	for b.Loop() {
+		client.TrackEvent("A message")
+	}
+
+	client.Close(context.Background())
+}
+
+func TestEndToEnd(t *testing.T) {
+	synctest.Run(func() {
+		xmit, server := newTestClientServer(t)
+		defer server.Close()
+
+		// Set up server response
+		server.responseData = []byte(`{"itemsReceived":4, "itemsAccepted":4, "errors":[]}`)
+		server.responseHeaders["Content-type"] = "application/json"
+
+		client := &Client{InstrumentationKey: test_ikey, Endpoint: xmit.(*httpTransmitter).endpoint, HTTPClient: xmit.(*httpTransmitter).client}
+		defer client.Close(context.Background())
+
+		// Track directly off the client
+		client.TrackEvent("client-event")
+
+		// NOTE: A lot of this is covered elsewhere, so we won't duplicate
+		// *too* much.
+
+		// Wait for automatic transmit -- get the request
+		slowTick(11)
+		req := server.waitForRequest(t)
+
+		// GZIP magic number
+		if len(req.body) < 2 || req.body[0] != 0x1f || req.body[1] != 0x8b {
+			t.Fatal("Missing gzip magic number")
+		}
+
+		// Decompress
+		reader, err := gzip.NewReader(bytes.NewReader(req.body))
+		if err != nil {
+			t.Fatalf("Coudln't create gzip reader: %s", err.Error())
+		}
+
+		// Read payload
+		body, err := io.ReadAll(reader)
+		reader.Close()
+		if err != nil {
+			t.Fatalf("Couldn't read compressed data: %s", err.Error())
+		}
```

Comment:
Seems like a decent chunk of this could be shared with `telemetry/internal/appinsights/transmitter_test.go`, up to the GZIP magic number?